{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import of Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "os.chdir(r'C:\\Users\\Teresa\\OneDrive - University Of Cambridge\\St Johns Cambridge\\1-Part II Biochemistry\\Part II Project\\Files\\Randomer sequencing')\n",
    "\n",
    "import sys\n",
    "from Bio.Seq import Seq\n",
    "# def to read fasta files\n",
    "def readFasta(fastaFile):\n",
    "    fh = open(fastaFile, 'r')\n",
    "    for line in fh:\n",
    "        if line[0] == '>':\n",
    "            header = line.rstrip()[1:]\n",
    "            if sys.version_info[0] < 3:\n",
    "                seq = fh.next().rstrip()\n",
    "            else:\n",
    "                seq = fh.readline().rstrip()\n",
    "        yield [header, seq]\n",
    "    fh.close()\n",
    "\n",
    "dict_rounds = {}\n",
    "\n",
    "\n",
    "rounds_names = ['RZ5N','F201N', 'F205N', 'N17L', 'N73L', 'N73H', 'FN17L', 'FN73L','FN73H'] #paper\n",
    "\n",
    "count = {}\n",
    "for i in rounds_names:\n",
    "    dict_rounds[i]={}\n",
    "    count[i]=0\n",
    "    for line in readFasta(i+'.fasta'): # looks through each line of each file\n",
    "        count[i]+=1\n",
    "        if  line[1] not in dict_rounds[i]: #adds to dict if seq not there\n",
    "            dict_rounds[i][line[1]] = 1\n",
    "            \n",
    "        else:\n",
    "            dict_rounds[i][line[1]] +=1            "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_filt3_range_rounds = {} #only take sequences that are multiples of 3 within right range \n",
    "#except pools with PNK are not multiple of 3 but right range and above 8 nt\n",
    "dict_filt3_rest={} #all sequences that are not multiple of 3\n",
    "\n",
    "def is_integer_num(n):\n",
    "    if isinstance(n, int):\n",
    "        return True\n",
    "    if isinstance(n, float):\n",
    "        return n.is_integer()\n",
    "    return False\n",
    "\n",
    "count_range = {} #new counting dictionary to include different categories\n",
    "length_count ={} #counts number of sequences in pool at length all in range and filtered 3\n",
    "\n",
    "\n",
    "count_categories=['total','greater5len','filtered3_greater5len','in_range','in_range_filtered3', \n",
    "                  'fraction_greater5len_total',\n",
    "                  'fraction_filtered3_greater5len_total',\n",
    "                  'fraction_in_range_total',\n",
    "                  'fraction_in_range_filtered3_total']\n",
    "\n",
    "for x in rounds_names:\n",
    "    count_range[x] = dict.fromkeys([count_categories[y] for y in range(len(count_categories))],0)\n",
    "    \n",
    "for i in rounds_names:\n",
    "    dict_filt3_range_rounds[i]={}\n",
    "    length_count[i]={}\n",
    "    dict_filt3_rest[i]={}\n",
    "    for seq in dict_rounds[i]:\n",
    "        count_range[i]['total']+= dict_rounds[i][seq]\n",
    "        \n",
    "        if len(seq)>5: #sequences with length 6 or more\n",
    "            count_range[i]['greater5len']+=dict_rounds[i][seq]\n",
    "            \n",
    "            if  seq not in dict_filt3_range_rounds[i]:\n",
    "                if is_integer_num(len(seq)/3): #sequences with length multiple of 3\n",
    "                    count_range[i]['filtered3_greater5len']+=dict_rounds[i][seq]\n",
    "                    \n",
    "                    if i == 'N17L' or i =='N73L' or i =='FN17L' or i =='FN73L': #range 12-30 nt plus minus 6 (below 6 already filtered before)\n",
    "                        if len(seq) <36:\n",
    "\n",
    "                            dict_filt3_range_rounds[i][seq]=dict_rounds[i][seq]\n",
    "                            count_range[i]['in_range_filtered3'] += dict_rounds[i][seq]\n",
    "                            \n",
    "                            if len(seq) not in length_count[i]:\n",
    "                                length_count[i][len(seq)]=dict_rounds[i][seq]\n",
    "                            else:\n",
    "                                length_count[i][len(seq)]+=dict_rounds[i][seq]\n",
    "\n",
    "                    elif i == 'N73H' or i =='FN73H': #range 33-150 nt plus minus 6 \n",
    "                        #--> adjusted to lower end as most reads seem to be shorter in pool\n",
    "                        if len(seq) > 21 and len(seq) < 156:\n",
    "                            dict_filt3_range_rounds[i][seq]=dict_rounds[i][seq] \n",
    "                            count_range[i]['in_range_filtered3'] += dict_rounds[i][seq]\n",
    "                            \n",
    "                            if len(seq) not in length_count[i]:\n",
    "                                length_count[i][len(seq)]=dict_rounds[i][seq]\n",
    "                            else:\n",
    "                                length_count[i][len(seq)]+=dict_rounds[i][seq]\n",
    "\n",
    "                    elif i == 'FN847':\n",
    "                        if len(seq) > 15 and len(seq) < 33: #around 27 nt plus minus 6 \n",
    "                            #adjusted to lower end as most reads here seem to be shorter (see plot)\n",
    "                            dict_filt3_range_rounds[i][seq]=dict_rounds[i][seq]\n",
    "                            count_range[i]['in_range_filtered3'] += dict_rounds[i][seq]\n",
    "                            \n",
    "                            if len(seq) not in length_count[i]:\n",
    "                                length_count[i][len(seq)]=dict_rounds[i][seq]\n",
    "                            else:\n",
    "                                length_count[i][len(seq)]+=dict_rounds[i][seq]\n",
    "\n",
    "                    elif i == 'RZ1N' or i == 'RZ5N' or i== 'DG1N' or i == 'DG5N' or i == 'F201N' or i =='F205N':\n",
    "                        if len(seq) < 27: #6-21 nt plus minus 6\n",
    "                            dict_filt3_range_rounds[i][seq]=dict_rounds[i][seq]\n",
    "                            count_range[i]['in_range_filtered3'] += dict_rounds[i][seq]\n",
    "                            \n",
    "                            if len(seq) not in length_count[i]:\n",
    "                                length_count[i][len(seq)]=dict_rounds[i][seq]\n",
    "                            else:\n",
    "                                length_count[i][len(seq)]+=dict_rounds[i][seq]\n",
    "\n",
    "                    elif i == 'N145' or i == 'FN145':\n",
    "                        if len(seq) < 51: #9-45 nt plus minus 6\n",
    "                            dict_filt3_range_rounds[i][seq]=dict_rounds[i][seq]\n",
    "                            count_range[i]['in_range_filtered3'] += dict_rounds[i][seq]\n",
    "                            \n",
    "                            if len(seq) not in length_count[i]:\n",
    "                                length_count[i][len(seq)]=dict_rounds[i][seq]\n",
    "                            else:\n",
    "                                length_count[i][len(seq)]+=dict_rounds[i][seq]\n",
    "                                \n",
    "                    elif i == 'N73A' or i == 'FN73A': #range 12-150 nt plus minus 6\n",
    "                        if len(seq) <156:\n",
    "                            dict_filt3_range_rounds[i][seq]=dict_rounds[i][seq] \n",
    "                            count_range[i]['in_range_filtered3'] += dict_rounds[i][seq]\n",
    "                            count_range[i]['in_range'] += dict_rounds[i][seq]\n",
    "                            \n",
    "                            if len(seq) not in length_count[i]:\n",
    "                                length_count[i][len(seq)]=dict_rounds[i][seq]\n",
    "                            else:\n",
    "                                length_count[i][len(seq)]+=dict_rounds[i][seq]\n",
    "                else: #takes additional sequences for these pools without requiring multiple of 3\n",
    "                    if i == 'N73A' or i == 'FN73A': #range 12-150 nt plus minus 6\n",
    "                        if len(seq) <156:\n",
    "                            dict_filt3_range_rounds[i][seq]=dict_rounds[i][seq] \n",
    "                            count_range[i]['in_range'] += dict_rounds[i][seq]\n",
    "                            \n",
    "                            if len(seq) not in length_count[i]:\n",
    "                                length_count[i][len(seq)]=dict_rounds[i][seq]\n",
    "                            else:\n",
    "                                length_count[i][len(seq)]+=dict_rounds[i][seq]\n",
    "                    if not is_integer_num(len(seq)/3):\n",
    "                        \n",
    "                        if i == 'N17L' or i =='N73L' or i =='FN17L' or i =='FN73L': #range 12-30 nt plus minus 6 (below 6 already filtered before)\n",
    "                            if len(seq) <36:\n",
    "\n",
    "                                dict_filt3_rest[i][seq]=dict_rounds[i][seq]\n",
    "\n",
    "                                \n",
    "\n",
    "                        elif i == 'N73H' or i =='FN73H': #range 33-150 nt plus minus 6 \n",
    "                            #--> adjusted to lower end as most reads seem to be shorter in pool\n",
    "                            if len(seq) > 21 and len(seq) < 156:\n",
    "                                dict_filt3_rest[i][seq]=dict_rounds[i][seq] \n",
    "\n",
    "                                \n",
    "                        elif i == 'RZ1N' or i == 'RZ5N' or i== 'DG1N' or i == 'DG5N' or i == 'F201N' or i =='F205N':\n",
    "                            if len(seq) < 27: #6-21 nt plus minus 6\n",
    "                                dict_filt3_rest[i][seq]=dict_rounds[i][seq]\n",
    "\n",
    "                            \n",
    "                                                        \n",
    "for i in rounds_names:\n",
    "    count_range[i]['fraction_greater5len_total'] = count_range[i]['greater5len']/count_range[i]['total']\n",
    "    count_range[i]['fraction_filtered3_greater5len_total']=count_range[i]['filtered3_greater5len']/count_range[i]['total']\n",
    "    count_range[i]['fraction_in_range_total'] = count_range[i]['in_range']/count_range[i]['total']\n",
    "    count_range[i]['fraction_in_range_filtered3_total']=count_range[i]['in_range_filtered3']/count_range[i]['total']\n",
    "    "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignments"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#takes very long to run\n",
    "type5 = 'GGATCTTCTCGATCTAACAAAAAAGACAAATCTGCCACAAAGCTTGAGAGCATCTTCGGATGCAGAGGCGGCAGCCTTCGGTGGCGCGATAGCGCCAACGTTCTCAACTATGACACGCAAAACGCGTGCTCCGTTGAATGGAGTTTATCATG'\n",
    "type5_reversecomp = 'CATGATAAACTCCATTCAACGGAGCACGCGTTTTGCGTGTCATAGTTGAGAACGTTGGCGCTATCGCGCCACCGAAGGCTGCCGCCTCTGCATCCGAAGATGCTCTCAAGCTTTGTGGCAGATTTGTCTTTTTTGTTAGATCGAGAAGATCC'\n",
    "type1 = 'GACCAATCTGCCCTCAGAGCCCGAGAACATCTTCGGATGCAGAGGAGGCAGGCTTCGGTGGCGCGATAGCGCCAACGTCCTCAACCTCCAATGCATCCCACCACATGATGAGCCTGAAGAGCCTTGGTTTTTTTG'\n",
    "type1_reversecomp = 'CAAAAAAACCAAGGCTCTTCAGGCTCATCATGTGGTGGGATGCATTGGAGGTTGAGGACGTTGGCGCTATCGCGCCACCGAAGCCTGCCTCCTCTGCATCCGAAGATGTTCTCGGGCTCTGAGGGCAGATTGGTC'\n",
    "\n",
    "ribozyme = [type5, type5_reversecomp, type1, type1_reversecomp]\n",
    "ribozyme_names = ['type5', 'type5_reversecomp', 'type1', 'type1_reversecomp']\n",
    "\n",
    "\n",
    "dict_alignment = {} # structure: pool --> ribozyme --> position --> score:/sequence:\n",
    "dict_not_ribozyme = {}\n",
    "dict_ribozyme={}\n",
    "dict_grey_area={} #not ribozyme and not not ribozyme --> in between --> not assigned\n",
    "alignment_count={}\n",
    "dict_alignment_noplusandminus = {}\n",
    "\n",
    "\n",
    "from Bio import pairwise2\n",
    "\n",
    "\n",
    "pool_selection = [\"RZ5N\",\"F201N\",\"F205N\",\"N17L\",\"N73L\",\"N73H\", \"FN17L\",\"FN73L\",\"FN73H\"]\n",
    "\n",
    "\n",
    "\n",
    "for i in pool_selection:\n",
    "    count = 0 # counts total number of sequences in pool\n",
    "    dict_alignment[i]={}\n",
    "    help_dict_align={} #temperorary dictionary \n",
    "    alignment_start={} #temporary dictionary\n",
    "    alignment_end={} #temporary dictionary\n",
    "    \n",
    "    #dictionaries collecting output sequences\n",
    "    dict_ribozyme[i]={}\n",
    "    dict_grey_area[i]={}\n",
    "    dict_not_ribozyme[i]={}\n",
    "    \n",
    "    #collects counts of sequences in each pool\n",
    "    alignment_count[i]={}\n",
    "    alignment_count[i]['aligned']=0\n",
    "    alignment_count[i]['grey_area']=0\n",
    "    alignment_count[i]['not_ribozyme']=0\n",
    "    alignment_count[i]['plus_aligned']=0\n",
    "    alignment_count[i]['minus_aligned']=0\n",
    "    alignment_count[i]['plus_and_minus']=0\n",
    "    alignment_count[i]['fraction_plus_aligned']=0\n",
    "    alignment_count[i]['fraction_minus_aligned'] = 0\n",
    "    alignment_count[i]['fraction_plus_and_minus'] =0\n",
    "    \n",
    "   \n",
    "    dict_alignment_noplusandminus[i] = {}\n",
    "    \n",
    "    for r in range(len(ribozyme)):\n",
    "        r_name = ribozyme_names[r]\n",
    "        dict_alignment[i][r_name]={} #add subdictionary\n",
    "        dict_alignment_noplusandminus[i][r_name]={}\n",
    "\n",
    "        r_len = len(ribozyme[r]) #length of ribozyme\n",
    "        positions_list = []\n",
    "        for z in range(r_len):\n",
    "            positions_list.append(z) #list of all positions ribozyme\n",
    "                \n",
    "        for p1 in positions_list: #prepare dictionary with all positions --> makes sure only record positions that exist as \n",
    "            #type 1 shorter than t5\n",
    "            dict_alignment[i][r_name][p1]={}\n",
    "            dict_alignment[i][r_name][p1]['score']=0\n",
    "            dict_alignment[i][r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]={}\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]['score']=0\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "        \n",
    "        dict_ribozyme[i][r_name]={}\n",
    "        \n",
    "    \n",
    "    for seq in dict_filt3_range_rounds[i]: #go through each file of filtered dictionary\n",
    "        #can insert length limitation here if only want to include length of sequences not filtered out previously\n",
    "        #change cutoff values here:\n",
    "        seq_len = len(seq)\n",
    "        cutoff_ribozyme = 0.66+2/seq_len\n",
    "        cutoff_not_ribozyme = 0.6\n",
    "        \n",
    "        count += dict_filt3_range_rounds[i][seq]\n",
    "        \n",
    "        aligned = 0\n",
    "        not_ribozyme = 0\n",
    "        alignment_start[seq]={}\n",
    "        alignment_end[seq]={}\n",
    "        help_dict_align[seq]={}\n",
    "        \n",
    "        for r1 in range(len(ribozyme)):\n",
    "            alignment = pairwise2.align.localms(ribozyme[r1],seq,1,0,-1000,-1000) #1 for match, 0 for mismatch, -1000 for gap\n",
    "            alignment_score = alignment[0][2]\n",
    "            alignment_score_normalized = alignment_score/seq_len\n",
    "                \n",
    "            if alignment_score_normalized <cutoff_not_ribozyme:\n",
    "                not_ribozyme +=1\n",
    "            \n",
    "            if alignment_score_normalized > cutoff_ribozyme: #cut off adjusted for length --> determined through looking at alignment distributions for different pools\n",
    "                aligned+=1\n",
    "                help_dict_align[seq][ribozyme_names[r1]] = dict_filt3_range_rounds[i][seq]\n",
    "                dict_ribozyme[i][ribozyme_names[r1]][seq]=dict_filt3_range_rounds[i][seq]\n",
    "                alignment_start[seq][ribozyme_names[r1]]=alignment[0][3] #retrieves start\n",
    "                alignment_end[seq][ribozyme_names[r1]]=alignment[0][4] #retrieves end\n",
    "                \n",
    "        if aligned == 0:\n",
    "            if not_ribozyme == 4:\n",
    "                alignment_count[i]['not_ribozyme']+= dict_filt3_range_rounds[i][seq]  \n",
    "                dict_not_ribozyme[i][seq]=dict_filt3_range_rounds[i][seq]\n",
    "            else:\n",
    "                alignment_count[i]['grey_area']+=dict_filt3_range_rounds[i][seq]\n",
    "                dict_grey_area[i][seq]=dict_filt3_range_rounds[i][seq]\n",
    "        else:\n",
    "            alignment_count[i]['aligned']+=dict_filt3_range_rounds[i][seq]\n",
    "            \n",
    "    for seq in help_dict_align: #write scores into alignment dictionary and triplet match\n",
    "        plus_strand = 0\n",
    "        minus_strand =0\n",
    "        \n",
    "        for r2 in range(len(ribozyme)):\n",
    "            if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                if (ribozyme_names[r2] == 'type5') or (ribozyme_names[r2] == 'type1'):\n",
    "                    plus_strand +=1\n",
    "                    \n",
    "                else:\n",
    "                    minus_strand +=1\n",
    "                    \n",
    "                my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                for y in range(my_range):\n",
    "                    position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "                    \n",
    "                    s = dict_filt3_range_rounds[i][seq]/(count*len(help_dict_align[seq]))\n",
    "                    dict_alignment[i][ribozyme_names[r2]][position]['score']+=s\n",
    "\n",
    "        if (plus_strand >0) and (minus_strand ==0):\n",
    "            alignment_count[i]['plus_aligned'] += dict_filt3_range_rounds[i][seq]\n",
    "        if (plus_strand >0) and (minus_strand >0):\n",
    "            alignment_count[i]['plus_and_minus'] += dict_filt3_range_rounds[i][seq]\n",
    "        if (plus_strand ==0) and (minus_strand >0):\n",
    "            alignment_count[i]['minus_aligned'] += dict_filt3_range_rounds[i][seq]\n",
    "            \n",
    "        if ((plus_strand == 0) and (minus_strand > 0)) or ((plus_strand > 0) and (minus_strand == 0)):\n",
    "            for r2 in range(len(ribozyme)):\n",
    "                if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                    \n",
    "                    my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                    for y in range(my_range):\n",
    "                        position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "\n",
    "                        s = dict_filt3_range_rounds[i][seq]/(count*len(help_dict_align[seq]))\n",
    "                        dict_alignment_noplusandminus[i][ribozyme_names[r2]][position]['score']+=s\n",
    "            \n",
    "            \n",
    "#turning dictionary of alignment to dataframe with mirrored complementary\n",
    "import pandas as pd\n",
    "rows = []\n",
    "\n",
    "r_order = ['1_type5', '3_type5_reversecomp', '2_type1', '4_type1_reversecomp']\n",
    "\n",
    "for pool in dict_alignment:\n",
    "    for r in range(len(dict_alignment[pool])):\n",
    "        for position in dict_alignment[pool][ribozyme_names[r]]:\n",
    "            data_row = {}\n",
    "            data_row['ribozyme']= r_order[r] \n",
    "            data_row['pool']=pool\n",
    "\n",
    "            if data_row['ribozyme']=='4_type1_reversecomp':\n",
    "                mirror_position = len(type1)-position-1\n",
    "                data_row['position']= mirror_position\n",
    "            elif data_row['ribozyme'] =='2_type5_reversecomp':\n",
    "                mirror_position = len(type5)-position-1\n",
    "                data_row['position']= mirror_position\n",
    "            else:\n",
    "                data_row['position']= position\n",
    "\n",
    "            data_row['sequence'] = dict_alignment[pool][ribozyme_names[r]][position]['sequence']\n",
    "            data_row['score']=dict_alignment[pool][ribozyme_names[r]][position]['score']\n",
    "\n",
    "            rows.append(data_row)\n",
    "\n",
    "df = pd.DataFrame(rows) \n",
    "df['position']=df['position']+1 #makes numbers more intuitive -> position 1 is 1 and not 0\n",
    "\n",
    "df['position_seq']=df['position'].astype(str)+'_'+df['sequence'].astype(str)\n",
    "df.to_csv('Ribozyme_align.csv')\n",
    "\n",
    "\n",
    "\n",
    "# saving dataframe of alignment count\n",
    "df = pd.DataFrame(alignment_count)\n",
    "\n",
    "df.index.name ='type'\n",
    "df.reset_index(inplace=True)\n",
    "df_long = pd.melt(df,id_vars=['type'],var_name='pool', value_name='value')\n",
    "\n",
    "\n",
    "df_long.to_csv(\"Alignment_count_all.csv\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Complementarity in Sequences from alignment all\n",
    "(need to run alignment all first)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_dict = {}\n",
    "all_dict['1-ribo']={}\n",
    "for pool in dict_ribozyme:\n",
    "    all_dict['1-ribo'][pool]={}\n",
    "    for ribo in dict_ribozyme[pool]:\n",
    "        for seq in dict_ribozyme[pool][ribo]:\n",
    "            all_dict['1-ribo'][pool][seq]=dict_ribozyme[pool][ribo][seq]\n",
    "all_dict['2-grey']=dict_grey_area\n",
    "all_dict['3-non_ribo']= dict_not_ribozyme\n",
    "\n",
    "#complementary sequences with more than 8 nucleotides\n",
    "import sys\n",
    "from Bio.Seq import Seq\n",
    "count_comp = {}\n",
    "\n",
    "for split in all_dict:\n",
    "    count_comp[split]={}\n",
    "    for i in all_dict[split]:\n",
    "        count_comp[split][i]={}\n",
    "        count_comp[split][i]['unique_complementary'] = 0\n",
    "        count_comp[split][i]['total_unique_seq']=0\n",
    "        count_comp[split][i]['complementary_with_repeat'] = 0\n",
    "        count_comp[split][i]['total_number_seq']=0\n",
    "        for seq in all_dict[split][i]:\n",
    "\n",
    "            if len(seq)>8:\n",
    "                count_comp[split][i]['total_unique_seq']+=1\n",
    "                count_comp[split][i]['total_number_seq'] +=dict_filt3_range_rounds[i][seq]\n",
    "                my_DNA = Seq(seq)\n",
    "                rev_complement = my_DNA.reverse_complement()\n",
    "                if rev_complement in dict_filt3_range_rounds[i]:\n",
    "                    count_comp[split][i]['unique_complementary']+=1\n",
    "                    count_comp[split][i]['complementary_with_repeat']+=dict_filt3_range_rounds[i][seq]\n",
    "\n",
    "\n",
    "        count_comp[split][i]['fraction_unique_complementary_seq']= count_comp[split][i]['unique_complementary']/count_comp[split][i]['total_unique_seq']\n",
    "        count_comp[split][i]['fraction_complementary_seq_with_repeat']= count_comp[split][i]['complementary_with_repeat']/count_comp[split][i]['total_number_seq']\n",
    "\n",
    "import pandas as pd\n",
    "rows = []\n",
    "\n",
    "for split in count_comp:\n",
    "    for pool in count_comp[split]:\n",
    "        for cat in count_comp[split][pool]:\n",
    "            \n",
    "            data_row = {}\n",
    "            data_row['split']= split\n",
    "            data_row['pool']=pool\n",
    "            data_row['category']=cat\n",
    "            data_row['value']=count_comp[split][pool][cat]\n",
    "            rows.append(data_row)\n",
    "\n",
    "df = pd.DataFrame(rows) \n",
    "\n",
    "df.to_csv('Complementarity_counts_splits.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uniqueness sequences from alignment all\n",
    "(need to run alignment all first)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_unique={}\n",
    "for split in all_dict:\n",
    "    count_unique[split]={}\n",
    "    for i in all_dict[split]:\n",
    "        unique = 0\n",
    "        total = 0\n",
    "        count_unique[split][i]={}\n",
    "\n",
    "        for seq in all_dict[split][i]:\n",
    "\n",
    "            if len(seq)>8:\n",
    "                unique +=1\n",
    "                total +=dict_filt3_range_rounds[i][seq]\n",
    "\n",
    "        count_unique[split][i]['fraction_unique']=unique/total    \n",
    "import pandas as pd\n",
    "rows = []\n",
    "\n",
    "for split in count_unique:\n",
    "    for pool in count_unique[split]:\n",
    "        for cat in count_unique[split][pool]:\n",
    "            \n",
    "            data_row = {}\n",
    "            data_row['split']= split\n",
    "            data_row['pool']=pool\n",
    "            data_row['category']=cat\n",
    "            data_row['value']=count_unique[split][pool][cat]\n",
    "            rows.append(data_row)\n",
    "\n",
    "df = pd.DataFrame(rows) \n",
    "\n",
    "df.to_csv('Uniqueness_counts_splits.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Nucleotide composition triplets non-ribozyme sequences\n",
    "(need to run alignment all first)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#looking at all sequences\n",
    "AT_seq={}\n",
    "\n",
    "pools =['F201N','F205N','FN17L','FN73L','FN73H','N17L']\n",
    "types =[\"3' triplet\",\"5' triplet\", \"All triplets in sequence\"]\n",
    "position = ['1st', '2nd', '3rd']\n",
    "nucleotide = ['A', 'T', 'G', 'C']\n",
    "\n",
    "list_1st = list(range(0,69,3))\n",
    "list_2nd = list(range(1,69,3))\n",
    "list_3rd = list(range(2,69,3))\n",
    "\n",
    "for i in pools:\n",
    "    AT_seq[i]={}\n",
    "    for t in types:\n",
    "        AT_seq[i][t]={}\n",
    "        for p in position:\n",
    "            AT_seq[i][t][p]={}\n",
    "            for n in nucleotide:\n",
    "                AT_seq[i][t][p][n]=0\n",
    "    \n",
    "    for seq in dict_not_ribozyme[i]:\n",
    "        len_s = len(seq)\n",
    "        for y in range(len_s):\n",
    "            if y in list_3rd:\n",
    "                AT_seq[i][types[2]]['3rd'][seq[y]]+=dict_not_ribozyme[i][seq]/(len_s/3*alignment_count[i]['not_ribozyme'])\n",
    "                if y == 2:\n",
    "                    AT_seq[i][types[1]]['3rd'][seq[y]]+=dict_not_ribozyme[i][seq]/alignment_count[i]['not_ribozyme']\n",
    "                if y == (len_s-1):\n",
    "                    AT_seq[i][types[0]]['3rd'][seq[y]]+=dict_not_ribozyme[i][seq]/alignment_count[i]['not_ribozyme']\n",
    "            elif y in list_2nd:\n",
    "                AT_seq[i][types[2]]['2nd'][seq[y]]+=dict_not_ribozyme[i][seq]/(len_s/3*alignment_count[i]['not_ribozyme'])\n",
    "                if y == 1:\n",
    "                    AT_seq[i][types[1]]['2nd'][seq[y]]+=dict_not_ribozyme[i][seq]/alignment_count[i]['not_ribozyme']\n",
    "                if y == (len_s-2):\n",
    "                    AT_seq[i][types[0]]['2nd'][seq[y]]+=dict_not_ribozyme[i][seq]/alignment_count[i]['not_ribozyme']\n",
    "            else: \n",
    "                AT_seq[i][types[2]]['1st'][seq[y]]+=dict_not_ribozyme[i][seq]/(len_s/3*alignment_count[i]['not_ribozyme'])\n",
    "                if y == 0:\n",
    "                    AT_seq[i][types[1]]['1st'][seq[y]]+=dict_not_ribozyme[i][seq]/alignment_count[i]['not_ribozyme']\n",
    "                if y == (len_s-3):\n",
    "                    AT_seq[i][types[0]]['1st'][seq[y]]+=dict_not_ribozyme[i][seq]/alignment_count[i]['not_ribozyme']\n",
    "\n",
    "AT_seq_df = {}\n",
    "AT_seq_df['Pool']=[]\n",
    "AT_seq_df['Types']=[]\n",
    "AT_seq_df['Position']=[]\n",
    "AT_seq_df['Nucleotide']=[]\n",
    "AT_seq_df['Value']=[]\n",
    "for i in AT_seq:\n",
    "    for t in AT_seq[i]:\n",
    "        for p in AT_seq[i][t]:\n",
    "            for n in AT_seq[i][t][p]:\n",
    "                AT_seq_df['Pool'].append(i)\n",
    "                AT_seq_df['Types'].append(t)\n",
    "                AT_seq_df['Position'].append(p)\n",
    "                AT_seq_df['Nucleotide'].append(n)\n",
    "                AT_seq_df['Value'].append(AT_seq[i][t][p][n])\n",
    "df = pd.DataFrame.from_dict(AT_seq_df)\n",
    "df.to_csv('Nucleotides_triplets.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment 9-nt longer sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#testing differences in long and short pools by taking random 9-nucleotide long stretches and comparing alignment frequency\n",
    "#takes very long to run\n",
    "type5 = 'GGATCTTCTCGATCTAACAAAAAAGACAAATCTGCCACAAAGCTTGAGAGCATCTTCGGATGCAGAGGCGGCAGCCTTCGGTGGCGCGATAGCGCCAACGTTCTCAACTATGACACGCAAAACGCGTGCTCCGTTGAATGGAGTTTATCATG'\n",
    "type5_reversecomp = 'CATGATAAACTCCATTCAACGGAGCACGCGTTTTGCGTGTCATAGTTGAGAACGTTGGCGCTATCGCGCCACCGAAGGCTGCCGCCTCTGCATCCGAAGATGCTCTCAAGCTTTGTGGCAGATTTGTCTTTTTTGTTAGATCGAGAAGATCC'\n",
    "type1 = 'GACCAATCTGCCCTCAGAGCCCGAGAACATCTTCGGATGCAGAGGAGGCAGGCTTCGGTGGCGCGATAGCGCCAACGTCCTCAACCTCCAATGCATCCCACCACATGATGAGCCTGAAGAGCCTTGGTTTTTTTG'\n",
    "type1_reversecomp = 'CAAAAAAACCAAGGCTCTTCAGGCTCATCATGTGGTGGGATGCATTGGAGGTTGAGGACGTTGGCGCTATCGCGCCACCGAAGCCTGCCTCCTCTGCATCCGAAGATGTTCTCGGGCTCTGAGGGCAGATTGGTC'\n",
    "\n",
    "ribozyme = [type5, type5_reversecomp, type1, type1_reversecomp]\n",
    "ribozyme_names = ['type5', 'type5_reversecomp', 'type1', 'type1_reversecomp']\n",
    "\n",
    "\n",
    "dict_alignment = {} # structure: pool --> ribozyme --> position --> score:/sequence:\n",
    "dict_not_ribozyme = {}\n",
    "dict_ribozyme={}\n",
    "dict_grey_area={} #not ribozyme and not not ribozyme --> in between --> not assigned\n",
    "alignment_count={}\n",
    "dict_alignment_noplusandminus={}\n",
    "\n",
    "\n",
    "from Bio import pairwise2\n",
    "\n",
    "pool_selection = [\"N73H\",\"FN73H\",\"N73L\",\"FN73L\"]\n",
    "\n",
    "\n",
    "\n",
    "for i in pool_selection:\n",
    "    count = 0 # counts total number of sequences in pool\n",
    "    dict_alignment[i]={}\n",
    "    help_dict_align={} #temperorary dictionary \n",
    "    alignment_start={} #temporary dictionary\n",
    "    alignment_end={} #temporary dictionary\n",
    "    \n",
    "    #dictionaries collecting output sequences\n",
    "    dict_ribozyme[i]={}\n",
    "    dict_grey_area[i]={}\n",
    "    dict_not_ribozyme[i]={}\n",
    "    \n",
    "    #collects counts of sequences in each pool\n",
    "    alignment_count[i]={}\n",
    "    alignment_count[i]['aligned']=0\n",
    "    alignment_count[i]['grey_area']=0\n",
    "    \n",
    "    alignment_count[i]['not_ribozyme']=0\n",
    "    alignment_count[i]['plus_aligned']=0\n",
    "    alignment_count[i]['minus_aligned']=0\n",
    "    alignment_count[i]['plus_and_minus']=0\n",
    "    alignment_count[i]['fraction_plus_aligned']=0\n",
    "    alignment_count[i]['fraction_minus_aligned'] = 0\n",
    "    alignment_count[i]['fraction_plus_and_minus'] =0\n",
    "    \n",
    "    dict_alignment_noplusandminus[i] = {}\n",
    "    \n",
    "    for r in range(len(ribozyme)):\n",
    "        r_name = ribozyme_names[r]\n",
    "        dict_alignment[i][r_name]={} #add subdictionary\n",
    "        dict_alignment_noplusandminus[i][r_name]={}    \n",
    "\n",
    "        r_len = len(ribozyme[r]) #length of ribozyme\n",
    "        positions_list = []\n",
    "        for z in range(r_len):\n",
    "            positions_list.append(z) #list of all positions ribozyme\n",
    "                \n",
    "        for p1 in positions_list: #prepare dictionary with all positions --> makes sure only record positions that exist as \n",
    "            #type 1 shorter than t5\n",
    "            dict_alignment[i][r_name][p1]={}\n",
    "            dict_alignment[i][r_name][p1]['score']=0\n",
    "            dict_alignment[i][r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]={}\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]['score']=0\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "        \n",
    "        dict_ribozyme[i][r_name]={}\n",
    "\n",
    "    \n",
    "    for seq in dict_filt3_range_rounds[i]: #go through each file of filtered dictionary\n",
    "        #can insert length limitation here if only want to include length of sequences not filtered out previously\n",
    "        #change cutoff values here:\n",
    "        seq_len = len(seq)\n",
    "        cutoff_ribozyme = 0.66+2/seq_len\n",
    "        cutoff_not_ribozyme = 0.6\n",
    "        \n",
    "        \n",
    "        \n",
    "        aligned = 0\n",
    "        not_ribozyme = 0\n",
    "        alignment_start[seq]={}\n",
    "        alignment_end[seq]={}\n",
    "        help_dict_align[seq]={}\n",
    "        \n",
    "        if seq_len >11:\n",
    "            seq_9 = seq[3:12]\n",
    "            count += dict_filt3_range_rounds[i][seq]\n",
    "        \n",
    "            for r1 in range(len(ribozyme)):\n",
    "                alignment = pairwise2.align.localms(ribozyme[r1],seq_9,1,0,-1000,-1000) #1 for match, 0 for mismatch, -1000 for gap\n",
    "                alignment_score = alignment[0][2]\n",
    "                alignment_score_normalized = alignment_score/9\n",
    "\n",
    "                if alignment_score_normalized <cutoff_not_ribozyme:\n",
    "                    not_ribozyme +=1\n",
    "\n",
    "                if alignment_score_normalized > cutoff_ribozyme: #cut off adjusted for length --> determined through looking at alignment distributions for different pools\n",
    "                    aligned+=1\n",
    "                    help_dict_align[seq][ribozyme_names[r1]] = dict_filt3_range_rounds[i][seq]\n",
    "                    dict_ribozyme[i][ribozyme_names[r1]][seq]=dict_filt3_range_rounds[i][seq]\n",
    "                    alignment_start[seq][ribozyme_names[r1]]=alignment[0][3] #retrieves start\n",
    "                    alignment_end[seq][ribozyme_names[r1]]=alignment[0][4] #retrieves end\n",
    "\n",
    "            if aligned == 0:\n",
    "                if not_ribozyme == 4:\n",
    "                    alignment_count[i]['not_ribozyme']+= dict_filt3_range_rounds[i][seq]  \n",
    "                    dict_not_ribozyme[i][seq]=dict_filt3_range_rounds[i][seq]\n",
    "                else:\n",
    "                    alignment_count[i]['grey_area']+=dict_filt3_range_rounds[i][seq]\n",
    "                    dict_grey_area[i][seq]=dict_filt3_range_rounds[i][seq]\n",
    "            else:\n",
    "                alignment_count[i]['aligned']+=dict_filt3_range_rounds[i][seq]\n",
    "            \n",
    "    for seq in help_dict_align: #write scores into alignment dictionary \n",
    "        plus_strand = 0\n",
    "        minus_strand =0\n",
    "        for r2 in range(len(ribozyme)):\n",
    "            if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                if (ribozyme_names[r2] == 'type5') or (ribozyme_names[r2] == 'type1'):\n",
    "                    plus_strand +=1\n",
    "                else:\n",
    "                    minus_strand +=1\n",
    "                my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                for y in range(my_range):\n",
    "                    position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "                    \n",
    "                    s = dict_filt3_range_rounds[i][seq]/(count*len(help_dict_align[seq]))\n",
    "                    dict_alignment[i][ribozyme_names[r2]][position]['score']+=s\n",
    "\n",
    "        if (plus_strand >0) and (minus_strand ==0):\n",
    "            alignment_count[i]['plus_aligned'] += dict_filt3_range_rounds[i][seq]\n",
    "        if (plus_strand >0) and (minus_strand >0):\n",
    "            alignment_count[i]['plus_and_minus'] += dict_filt3_range_rounds[i][seq]\n",
    "        if (plus_strand ==0) and (minus_strand >0):\n",
    "            alignment_count[i]['minus_aligned'] += dict_filt3_range_rounds[i][seq]\n",
    "        \n",
    "        if ((plus_strand == 0) and (minus_strand > 0)) or ((plus_strand > 0) and (minus_strand == 0)):\n",
    "            for r2 in range(len(ribozyme)):\n",
    "                if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                    \n",
    "                    my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                    for y in range(my_range):\n",
    "                        position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "\n",
    "                        s = dict_filt3_range_rounds[i][seq]/(count*len(help_dict_align[seq]))\n",
    "                        dict_alignment_noplusandminus[i][ribozyme_names[r2]][position]['score']+=s\n",
    "        \n",
    "            \n",
    "    alignment_count[i]['fraction_plus_aligned']=alignment_count[i]['plus_aligned']/alignment_count[i]['aligned']\n",
    "    alignment_count[i]['fraction_minus_aligned'] = alignment_count[i]['minus_aligned']/alignment_count[i]['aligned']\n",
    "    alignment_count[i]['fraction_plus_and_minus'] = alignment_count[i]['plus_and_minus']/alignment_count[i]['aligned']\n",
    "\n",
    "\n",
    "# saving dataframe of alignment count\n",
    "df = pd.DataFrame(alignment_count)\n",
    "\n",
    "df.index.name ='type'\n",
    "df.reset_index(inplace=True)\n",
    "df_long = pd.melt(df,id_vars=['type'],var_name='pool', value_name='value')\n",
    "\n",
    "\n",
    "df_long.to_csv(\"Alignment_count_longer_9nt.csv\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random 9-nt sequences with same composition alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Random alignment with same triplet composition\n",
    "\n",
    "import numpy as np\n",
    "pool_selection = [\"N73H\",\"FN73H\",\"N73L\",\"FN73L\"]\n",
    "nucleotides = ['G', 'T', 'C', 'A']\n",
    "nucleotides2 = ['T','C','A','G']\n",
    "codon ={}\n",
    "dict_random = {}\n",
    "for i in pool_selection:\n",
    "    dict_random[i]={}\n",
    "    count = 0\n",
    "    for n1 in nucleotides:\n",
    "        for n2 in nucleotides2:\n",
    "            for n3 in nucleotides2:\n",
    "                codon[n1+n2+n3] = 0\n",
    "    for seq in dict_filt3_range_rounds[i]:\n",
    "        len_s =len(seq)\n",
    "        count += len_s/3*dict_filt3_range_rounds[i][seq]\n",
    "        for l in range(0,len_s-2,3):\n",
    "            n_1 = seq[l]\n",
    "            n_2 = seq[l+1]\n",
    "            n_3 = seq[l+2]\n",
    "            codon[n_1+n_2+n_3]+=dict_filt3_range_rounds[i][seq]\n",
    "    for n1 in nucleotides:\n",
    "        for n2 in nucleotides2:\n",
    "            for n3 in nucleotides2:\n",
    "                \n",
    "                codon[n1+n2+n3]=codon[n1+n2+n3]/count\n",
    "    codon_new={}\n",
    "    codon_new['Codon']=[]\n",
    "    codon_new['Frequency']=[]\n",
    "    for c in codon:\n",
    "        codon_new['Codon'].append(c)\n",
    "        codon_new['Frequency'].append(codon[c])\n",
    "    df = pd.DataFrame.from_dict(codon_new)\n",
    "    df.to_csv('Overall_Codon_Frequency_'+i+'.csv')\n",
    "    \n",
    "    for j in range(10000):\n",
    "        a=np.random.choice(codon_new['Codon'], p=codon_new['Frequency'],  size=1)\n",
    "        b=np.random.choice(codon_new['Codon'], p=codon_new['Frequency'],  size=1)\n",
    "        c=np.random.choice(codon_new['Codon'], p=codon_new['Frequency'],  size=1)\n",
    "        d = a[0]+b[0]+c[0]\n",
    "        if  d not in dict_random[i]: #adds to dict if seq not there\n",
    "            dict_random[i][d] = 1\n",
    "            \n",
    "        else:\n",
    "            dict_random[i][d] +=1\n",
    "\n",
    "#alignment of sequences from dict_random --> random sequences based on triplet frequency\n",
    "type5 = 'GGATCTTCTCGATCTAACAAAAAAGACAAATCTGCCACAAAGCTTGAGAGCATCTTCGGATGCAGAGGCGGCAGCCTTCGGTGGCGCGATAGCGCCAACGTTCTCAACTATGACACGCAAAACGCGTGCTCCGTTGAATGGAGTTTATCATG'\n",
    "type5_reversecomp = 'CATGATAAACTCCATTCAACGGAGCACGCGTTTTGCGTGTCATAGTTGAGAACGTTGGCGCTATCGCGCCACCGAAGGCTGCCGCCTCTGCATCCGAAGATGCTCTCAAGCTTTGTGGCAGATTTGTCTTTTTTGTTAGATCGAGAAGATCC'\n",
    "type1 = 'GACCAATCTGCCCTCAGAGCCCGAGAACATCTTCGGATGCAGAGGAGGCAGGCTTCGGTGGCGCGATAGCGCCAACGTCCTCAACCTCCAATGCATCCCACCACATGATGAGCCTGAAGAGCCTTGGTTTTTTTG'\n",
    "type1_reversecomp = 'CAAAAAAACCAAGGCTCTTCAGGCTCATCATGTGGTGGGATGCATTGGAGGTTGAGGACGTTGGCGCTATCGCGCCACCGAAGCCTGCCTCCTCTGCATCCGAAGATGTTCTCGGGCTCTGAGGGCAGATTGGTC'\n",
    "\n",
    "ribozyme = [type5, type5_reversecomp, type1, type1_reversecomp]\n",
    "ribozyme_names = ['type5', 'type5_reversecomp', 'type1', 'type1_reversecomp']\n",
    "\n",
    "\n",
    "dict_alignment = {} # structure: pool --> ribozyme --> position --> score:/sequence:\n",
    "dict_not_ribozyme = {}\n",
    "dict_ribozyme={}\n",
    "dict_grey_area={} #not ribozyme and not not ribozyme --> in between --> not assigned\n",
    "alignment_count={}\n",
    "dict_alignment_noplusandminus={}\n",
    "\n",
    "\n",
    "from Bio import pairwise2\n",
    "\n",
    "pool_selection = [\"N73H\",\"FN73H\",\"N73L\",\"FN73L\"]\n",
    "    #\"RZ5N\",\"F201N\",\"F205N\",\"N17L\",\"N73L\",\"N73H\", \"FN17L\",\"FN73L\",\"FN73H\"]\n",
    "\n",
    "\n",
    "\n",
    "for i in pool_selection:\n",
    "    count = 0 # counts total number of sequences in pool\n",
    "    dict_alignment[i]={}\n",
    "    help_dict_align={} #temperorary dictionary \n",
    "    alignment_start={} #temporary dictionary\n",
    "    alignment_end={} #temporary dictionary\n",
    "    \n",
    "    #dictionaries collecting output sequences\n",
    "    dict_ribozyme[i]={}\n",
    "    dict_grey_area[i]={}\n",
    "    dict_not_ribozyme[i]={}\n",
    "    \n",
    "    #collects counts of sequences in each pool\n",
    "    alignment_count[i]={}\n",
    "    alignment_count[i]['aligned']=0\n",
    "    alignment_count[i]['grey_area']=0\n",
    "    \n",
    "    alignment_count[i]['not_ribozyme']=0\n",
    "    alignment_count[i]['plus_aligned']=0\n",
    "    alignment_count[i]['minus_aligned']=0\n",
    "    alignment_count[i]['plus_and_minus']=0\n",
    "    alignment_count[i]['fraction_plus_aligned']=0\n",
    "    alignment_count[i]['fraction_minus_aligned'] = 0\n",
    "    alignment_count[i]['fraction_plus_and_minus'] =0\n",
    "    \n",
    "    dict_alignment_noplusandminus[i] = {}\n",
    "    \n",
    "    for r in range(len(ribozyme)):\n",
    "        r_name = ribozyme_names[r]\n",
    "        dict_alignment[i][r_name]={} #add subdictionary\n",
    "        dict_alignment_noplusandminus[i][r_name]={}    \n",
    "\n",
    "        r_len = len(ribozyme[r]) #length of ribozyme\n",
    "        positions_list = []\n",
    "        for z in range(r_len):\n",
    "            positions_list.append(z) #list of all positions ribozyme\n",
    "                \n",
    "        for p1 in positions_list: #prepare dictionary with all positions --> makes sure only record positions that exist as \n",
    "            #type 1 shorter than t5\n",
    "            dict_alignment[i][r_name][p1]={}\n",
    "            dict_alignment[i][r_name][p1]['score']=0\n",
    "            dict_alignment[i][r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]={}\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]['score']=0\n",
    "            dict_alignment_noplusandminus[i][r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "        \n",
    "        dict_ribozyme[i][r_name]={}\n",
    "\n",
    "    \n",
    "    for seq in dict_random[i]: #go through each file of filtered dictionary\n",
    "        #can insert length limitation here if only want to include length of sequences not filtered out previously\n",
    "        #change cutoff values here:\n",
    "        seq_len = len(seq)\n",
    "        cutoff_ribozyme = 0.66+2/seq_len\n",
    "        cutoff_not_ribozyme = 0.6\n",
    "        \n",
    "        \n",
    "        \n",
    "        aligned = 0\n",
    "        not_ribozyme = 0\n",
    "        alignment_start[seq]={}\n",
    "        alignment_end[seq]={}\n",
    "        help_dict_align[seq]={}\n",
    "        \n",
    "        \n",
    "        count += dict_random[i][seq]\n",
    "\n",
    "        for r1 in range(len(ribozyme)):\n",
    "            alignment = pairwise2.align.localms(ribozyme[r1],seq,1,0,-1000,-1000) #1 for match, 0 for mismatch, -1000 for gap\n",
    "            alignment_score = alignment[0][2]\n",
    "            alignment_score_normalized = alignment_score/9\n",
    "\n",
    "            if alignment_score_normalized <cutoff_not_ribozyme:\n",
    "                not_ribozyme +=1\n",
    "\n",
    "            if alignment_score_normalized > cutoff_ribozyme: #cut off adjusted for length --> determined through looking at alignment distributions for different pools\n",
    "                aligned+=1\n",
    "                help_dict_align[seq][ribozyme_names[r1]] = dict_random[i][seq]\n",
    "                dict_ribozyme[i][ribozyme_names[r1]][seq]=dict_random[i][seq]\n",
    "                alignment_start[seq][ribozyme_names[r1]]=alignment[0][3] #retrieves start\n",
    "                alignment_end[seq][ribozyme_names[r1]]=alignment[0][4] #retrieves end\n",
    "\n",
    "        if aligned == 0:\n",
    "            if not_ribozyme == 4:\n",
    "                alignment_count[i]['not_ribozyme']+= dict_random[i][seq]  \n",
    "                dict_not_ribozyme[i][seq]=dict_random[i][seq]\n",
    "            else:\n",
    "                alignment_count[i]['grey_area']+=dict_random[i][seq]\n",
    "                dict_grey_area[i][seq]=dict_random[i][seq]\n",
    "        else:\n",
    "            alignment_count[i]['aligned']+=dict_random[i][seq]\n",
    "            \n",
    "    for seq in help_dict_align: #write scores into alignment dictionary \n",
    "        plus_strand = 0\n",
    "        minus_strand =0\n",
    "        for r2 in range(len(ribozyme)):\n",
    "            if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                if (ribozyme_names[r2] == 'type5') or (ribozyme_names[r2] == 'type1'):\n",
    "                    plus_strand +=1\n",
    "                else:\n",
    "                    minus_strand +=1\n",
    "                my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                for y in range(my_range):\n",
    "                    position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "                    \n",
    "                    s = dict_random[i][seq]/(count*len(help_dict_align[seq]))\n",
    "                    dict_alignment[i][ribozyme_names[r2]][position]['score']+=s\n",
    "\n",
    "        if (plus_strand >0) and (minus_strand ==0):\n",
    "            alignment_count[i]['plus_aligned'] += dict_random[i][seq]\n",
    "        if (plus_strand >0) and (minus_strand >0):\n",
    "            alignment_count[i]['plus_and_minus'] += dict_random[i][seq]\n",
    "        if (plus_strand ==0) and (minus_strand >0):\n",
    "            alignment_count[i]['minus_aligned'] += dict_random[i][seq]\n",
    "        \n",
    "        if ((plus_strand == 0) and (minus_strand > 0)) or ((plus_strand > 0) and (minus_strand == 0)):\n",
    "            for r2 in range(len(ribozyme)):\n",
    "                if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                    \n",
    "                    my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                    for y in range(my_range):\n",
    "                        position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "\n",
    "                        s = dict_random[i][seq]/(count*len(help_dict_align[seq]))\n",
    "                        dict_alignment_noplusandminus[i][ribozyme_names[r2]][position]['score']+=s\n",
    "        \n",
    "            \n",
    "\n",
    "# saving dataframe of alignment count\n",
    "df = pd.DataFrame(alignment_count)\n",
    "\n",
    "df.index.name ='type'\n",
    "df.reset_index(inplace=True)\n",
    "df_long = pd.melt(df,id_vars=['type'],var_name='pool', value_name='value')\n",
    "\n",
    "\n",
    "df_long.to_csv(\"Alignment_count_random_9nt.csv\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment combined H and I pools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_selection = ['N73L','N73H'] # pool H and I\n",
    "\n",
    "dict_comb = {}\n",
    "for i in sub_selection:\n",
    "    for seq in dict_filt3_range_rounds[i]:\n",
    "        if seq not in dict_comb:\n",
    "            dict_comb[seq]=dict_filt3_range_rounds[i][seq]\n",
    "        else:\n",
    "            dict_comb[seq]+=dict_filt3_range_rounds[i][seq]\n",
    "\n",
    "\n",
    "type5 = 'GGATCTTCTCGATCTAACAAAAAAGACAAATCTGCCACAAAGCTTGAGAGCATCTTCGGATGCAGAGGCGGCAGCCTTCGGTGGCGCGATAGCGCCAACGTTCTCAACTATGACACGCAAAACGCGTGCTCCGTTGAATGGAGTTTATCATG'\n",
    "type5_reversecomp = 'CATGATAAACTCCATTCAACGGAGCACGCGTTTTGCGTGTCATAGTTGAGAACGTTGGCGCTATCGCGCCACCGAAGGCTGCCGCCTCTGCATCCGAAGATGCTCTCAAGCTTTGTGGCAGATTTGTCTTTTTTGTTAGATCGAGAAGATCC'\n",
    "type1 = 'GACCAATCTGCCCTCAGAGCCCGAGAACATCTTCGGATGCAGAGGAGGCAGGCTTCGGTGGCGCGATAGCGCCAACGTCCTCAACCTCCAATGCATCCCACCACATGATGAGCCTGAAGAGCCTTGGTTTTTTTG'\n",
    "type1_reversecomp = 'CAAAAAAACCAAGGCTCTTCAGGCTCATCATGTGGTGGGATGCATTGGAGGTTGAGGACGTTGGCGCTATCGCGCCACCGAAGCCTGCCTCCTCTGCATCCGAAGATGTTCTCGGGCTCTGAGGGCAGATTGGTC'\n",
    "\n",
    "ribozyme = [type5, type5_reversecomp, type1, type1_reversecomp]\n",
    "ribozyme_names = ['type5', 'type5_reversecomp', 'type1', 'type1_reversecomp']\n",
    "\n",
    "\n",
    "dict_alignment = {} # structure: pool --> ribozyme --> position --> score:/sequence:\n",
    "dict_not_ribozyme = {}\n",
    "dict_ribozyme={}\n",
    "dict_grey_area={} #not ribozyme and not not ribozyme --> in between --> not assigned\n",
    "alignment_count={}\n",
    "dict_alignment_noplusandminus={}\n",
    "\n",
    "\n",
    "from Bio import pairwise2\n",
    "import datetime\n",
    "\n",
    "\n",
    "\n",
    "count = 0 # counts total number of sequences in pool\n",
    "dict_alignment={}\n",
    "help_dict_align={} #temperorary dictionary \n",
    "alignment_start={} #temporary dictionary\n",
    "alignment_end={} #temporary dictionary\n",
    "\n",
    "#dictionaries collecting output sequences\n",
    "dict_ribozyme={}\n",
    "dict_grey_area={}\n",
    "dict_not_ribozyme={}\n",
    "\n",
    "#collects counts of sequences in each pool\n",
    "alignment_count={}\n",
    "alignment_count['aligned']=0\n",
    "alignment_count['grey_area']=0\n",
    "alignment_count['not_ribozyme']=0\n",
    "alignment_count['plus_aligned']=0\n",
    "alignment_count['minus_aligned']=0\n",
    "alignment_count['plus_and_minus']=0\n",
    "alignment_count['fraction_plus_aligned']=0\n",
    "alignment_count['fraction_minus_aligned'] = 0\n",
    "alignment_count['fraction_plus_and_minus'] =0\n",
    "\n",
    "dict_alignment_noplusandminus = {}\n",
    "\n",
    "list_len_seq =[]\n",
    "list_score=[]\n",
    "list_score_norm=[]\n",
    "classification =[]\n",
    "sequences =[]\n",
    "\n",
    "track =0\n",
    "track2=0\n",
    "    \n",
    "for r in range(len(ribozyme)):\n",
    "    r_name = ribozyme_names[r]\n",
    "    dict_alignment[r_name]={} #add subdictionary\n",
    "    dict_alignment_noplusandminus[r_name]={}    \n",
    "\n",
    "    r_len = len(ribozyme[r]) #length of ribozyme\n",
    "    positions_list = []\n",
    "    for z in range(r_len):\n",
    "        positions_list.append(z) #list of all positions ribozyme\n",
    "            \n",
    "    for p1 in positions_list: #prepare dictionary with all positions --> makes sure only record positions that exist as \n",
    "        #type 1 shorter than t5\n",
    "        dict_alignment[r_name][p1]={}\n",
    "        dict_alignment[r_name][p1]['score']=0\n",
    "        dict_alignment[r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "        dict_alignment_noplusandminus[r_name][p1]={}\n",
    "        dict_alignment_noplusandminus[r_name][p1]['score']=0\n",
    "        dict_alignment_noplusandminus[r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "    \n",
    "    dict_ribozyme[r_name]={}\n",
    "\n",
    "    \n",
    "for seq in dict_comb: #go through each file of filtered dictionary\n",
    "    #can insert length limitation here if only want to include length of sequences not filtered out previously\n",
    "    #change cutoff values here:\n",
    "    track2+=1\n",
    "    seq_len = len(seq)\n",
    "    #append seq_len to list_len_seq\n",
    "    list_len_seq.append(seq_len)\n",
    "    sequences.append(seq)\n",
    "\n",
    "    cutoff_ribozyme = 0.66+2/seq_len\n",
    "    cutoff_not_ribozyme = 0.6\n",
    "    \n",
    "    aligned = 0\n",
    "    not_ribozyme = 0\n",
    "    alignment_start[seq]={}\n",
    "    alignment_end[seq]={}\n",
    "    help_dict_align[seq]={}\n",
    "    \n",
    "        \n",
    "    count += dict_comb[seq]\n",
    "\n",
    "    for r1 in range(len(ribozyme)):\n",
    "        alignment = pairwise2.align.localms(ribozyme[r1],seq,1,0,-1000,-1000) #1 for match, 0 for mismatch, -1000 for gap\n",
    "        alignment_score = alignment[0][2]\n",
    "        list_score.append(alignment_score)\n",
    "\n",
    "        alignment_score_normalized = alignment_score/seq_len\n",
    "        list_score_norm.append(alignment_score_normalized)\n",
    "\n",
    "        if alignment_score_normalized <cutoff_not_ribozyme:\n",
    "            not_ribozyme +=1\n",
    "\n",
    "        if alignment_score_normalized > cutoff_ribozyme: #cut off adjusted for length --> determined through looking at alignment distributions for different pools\n",
    "            aligned+=1\n",
    "            help_dict_align[seq][ribozyme_names[r1]] = dict_comb[seq]\n",
    "            dict_ribozyme[ribozyme_names[r1]][seq]=dict_comb[seq]\n",
    "            alignment_start[seq][ribozyme_names[r1]]=alignment[0][3] #retrieves start\n",
    "            alignment_end[seq][ribozyme_names[r1]]=alignment[0][4] #retrieves end\n",
    "\n",
    "    if aligned == 0:\n",
    "        if not_ribozyme == 4:\n",
    "            alignment_count['not_ribozyme']+= dict_comb[seq]  \n",
    "            classification.append('not_ribozyme')\n",
    "            dict_not_ribozyme[seq]=dict_comb[seq]\n",
    "        else:\n",
    "            alignment_count['grey_area']+=dict_comb[seq]\n",
    "            dict_grey_area[seq]=dict_comb[seq]\n",
    "            classification.append('grey_area')\n",
    "    else:\n",
    "        alignment_count['aligned']+=dict_comb[seq]\n",
    "        classification.append('aligned')\n",
    "\n",
    "    #if track2 divisible by 1000, print track2 and time\n",
    "    if track2%1000 == 0:\n",
    "        print(track2)\n",
    "        print(datetime.datetime.now())\n",
    "        \n",
    "for seq in help_dict_align: #write scores into alignment dictionary \n",
    "    \n",
    "    plus_strand = 0\n",
    "    minus_strand =0\n",
    "    for r2 in range(len(ribozyme)):\n",
    "        if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "            if (ribozyme_names[r2] == 'type5') or (ribozyme_names[r2] == 'type1'):\n",
    "                plus_strand +=1\n",
    "            else:\n",
    "                minus_strand +=1\n",
    "            my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "            for y in range(my_range):\n",
    "                position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "                \n",
    "                s = dict_comb[seq]/(count*len(help_dict_align[seq]))\n",
    "                dict_alignment[ribozyme_names[r2]][position]['score']+=s\n",
    "\n",
    "    if (plus_strand >0) and (minus_strand ==0):\n",
    "        alignment_count['plus_aligned'] += dict_comb[seq]\n",
    "        #change value in classification list at same position as seq\n",
    "        classification[track]='plus_aligned'\n",
    "\n",
    "    if (plus_strand >0) and (minus_strand >0):\n",
    "        alignment_count['plus_and_minus'] += dict_comb[seq]\n",
    "        classification[track]='plus_and_minus'\n",
    "    if (plus_strand ==0) and (minus_strand >0):\n",
    "        alignment_count['minus_aligned'] += dict_comb[seq]\n",
    "        classification[track]='minus_aligned'\n",
    "    \n",
    "    if ((plus_strand == 0) and (minus_strand > 0)) or ((plus_strand > 0) and (minus_strand == 0)):\n",
    "        for r2 in range(len(ribozyme)):\n",
    "            if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                    \n",
    "                my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                for y in range(my_range):\n",
    "                    position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "\n",
    "                    s = dict_comb[seq]/(count*len(help_dict_align[seq])) #len(help_dict_align[seq] is number of ribozyme sequences aligned to\n",
    "                    dict_alignment_noplusandminus[ribozyme_names[r2]][position]['score']+=s\n",
    "\n",
    "    track +=1\n",
    "    \n",
    "        \n",
    "alignment_count['fraction_plus_aligned']=alignment_count['plus_aligned']/alignment_count['aligned']\n",
    "alignment_count['fraction_minus_aligned'] = alignment_count['minus_aligned']/alignment_count['aligned']\n",
    "alignment_count['fraction_plus_and_minus'] = alignment_count['plus_and_minus']/alignment_count['aligned']\n",
    "#turning dictionary of alignment to dataframe with mirrored complementary\n",
    "import pandas as pd\n",
    "rows = []\n",
    "\n",
    "r_order = ['1_type5', '3_type5_reversecomp', '2_type1', '4_type1_reversecomp']\n",
    "\n",
    "\n",
    "#make dataframe with lists sequences, list_len_seq, list_score, list_score_norm, classification as columns\n",
    "list_score_norm_max = []\n",
    "list_score_max=[]\n",
    "for i in range(0, len(list_score_norm), 4):\n",
    "    list_score_norm_max.append(max(list_score_norm[i:i+4]))\n",
    "    list_score_max.append(max(list_score[i:i+4]))\n",
    "\n",
    "df=pd.DataFrame({'sequences':sequences, 'len_seq':list_len_seq, 'score':list_score_max, 'score_norm':list_score_norm_max, 'classification':classification})\n",
    "\n",
    "\n",
    "#make column class where class is classification if \"not_ribozyme\" or \"grey_area\" and is 'aligned' for \"plus_aligned\",  \"minus_aligned\",\"plus_and_minus\"\n",
    "df['class'] = df['classification']\n",
    "df.loc[df['classification']=='plus_aligned', 'class'] = 'aligned'\n",
    "df.loc[df['classification']=='minus_aligned', 'class'] = 'aligned'\n",
    "df.loc[df['classification']=='plus_and_minus', 'class'] = 'aligned'\n",
    "\n",
    "df.to_csv(\"Length_vs_alignment_score_H_I.csv\",index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment random Sequences with Triplet composition as pool H and I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# random sequences based on codon composition in dataset\n",
    "import numpy as np\n",
    "nucleotides = ['G', 'T', 'C', 'A']\n",
    "nucleotides2 = ['T','C','A','G']\n",
    "codon ={}\n",
    "dict_random = {}\n",
    "len_dict ={}\n",
    "    \n",
    "count = 0\n",
    "total = sum(dict_comb.values())\n",
    "for n1 in nucleotides:\n",
    "    for n2 in nucleotides2:\n",
    "        for n3 in nucleotides2:\n",
    "            codon[n1+n2+n3] = 0\n",
    "for seq in dict_comb:\n",
    "    len_s =len(seq)\n",
    "    if len_s in len_dict:\n",
    "        len_dict[len_s]+=dict_comb[seq]/total\n",
    "    else:\n",
    "        len_dict[len_s]=dict_comb[seq]/total\n",
    "    count += len_s/3*dict_comb[seq]\n",
    "    for l in range(0,len_s-2,3):\n",
    "        n_1 = seq[l]\n",
    "        n_2 = seq[l+1]\n",
    "        n_3 = seq[l+2]\n",
    "        codon[n_1+n_2+n_3]+=dict_comb[seq]\n",
    "for n1 in nucleotides:\n",
    "    for n2 in nucleotides2:\n",
    "        for n3 in nucleotides2:\n",
    "            \n",
    "            codon[n1+n2+n3]=codon[n1+n2+n3]/count\n",
    "codon_new={}\n",
    "codon_new['Codon']=[]\n",
    "codon_new['Frequency']=[]\n",
    "for c in codon:\n",
    "    codon_new['Codon'].append(c)\n",
    "    codon_new['Frequency'].append(codon[c])\n",
    "\n",
    "#list of keys in len_dict\n",
    "len_dict_keys = list(len_dict.keys())\n",
    "#list of values in len_dict\n",
    "len_dict_values = list(len_dict.values())\n",
    "\n",
    "count =0\n",
    "    \n",
    "while count < len(dict_comb):\n",
    "    length = np.random.choice(len_dict_keys, p=len_dict_values, size=1)\n",
    "    codon_number = length/3\n",
    "    sequence =''\n",
    "    for c in range(int(codon_number)):\n",
    "        codon = np.random.choice(codon_new['Codon'], p=codon_new['Frequency'],  size=1)\n",
    "        sequence += codon[0]\n",
    "    \n",
    "    if  sequence not in dict_random: #adds to dict if seq not there\n",
    "        dict_random[sequence] = 1\n",
    "        count +=1\n",
    "        \n",
    "    else:\n",
    "        dict_random[sequence] +=1\n",
    "        \n",
    "\n",
    "\n",
    "type5 = 'GGATCTTCTCGATCTAACAAAAAAGACAAATCTGCCACAAAGCTTGAGAGCATCTTCGGATGCAGAGGCGGCAGCCTTCGGTGGCGCGATAGCGCCAACGTTCTCAACTATGACACGCAAAACGCGTGCTCCGTTGAATGGAGTTTATCATG'\n",
    "type5_reversecomp = 'CATGATAAACTCCATTCAACGGAGCACGCGTTTTGCGTGTCATAGTTGAGAACGTTGGCGCTATCGCGCCACCGAAGGCTGCCGCCTCTGCATCCGAAGATGCTCTCAAGCTTTGTGGCAGATTTGTCTTTTTTGTTAGATCGAGAAGATCC'\n",
    "type1 = 'GACCAATCTGCCCTCAGAGCCCGAGAACATCTTCGGATGCAGAGGAGGCAGGCTTCGGTGGCGCGATAGCGCCAACGTCCTCAACCTCCAATGCATCCCACCACATGATGAGCCTGAAGAGCCTTGGTTTTTTTG'\n",
    "type1_reversecomp = 'CAAAAAAACCAAGGCTCTTCAGGCTCATCATGTGGTGGGATGCATTGGAGGTTGAGGACGTTGGCGCTATCGCGCCACCGAAGCCTGCCTCCTCTGCATCCGAAGATGTTCTCGGGCTCTGAGGGCAGATTGGTC'\n",
    "\n",
    "ribozyme = [type5, type5_reversecomp, type1, type1_reversecomp]\n",
    "ribozyme_names = ['type5', 'type5_reversecomp', 'type1', 'type1_reversecomp']\n",
    "\n",
    "\n",
    "dict_alignment = {} # structure: pool --> ribozyme --> position --> score:/sequence:\n",
    "dict_not_ribozyme = {}\n",
    "dict_ribozyme={}\n",
    "dict_grey_area={} #not ribozyme and not not ribozyme --> in between --> not assigned\n",
    "alignment_count={}\n",
    "dict_alignment_noplusandminus={}\n",
    "\n",
    "\n",
    "from Bio import pairwise2\n",
    "import datetime\n",
    "\n",
    "count = 0 # counts total number of sequences in pool\n",
    "dict_alignment={}\n",
    "help_dict_align={} #temperorary dictionary \n",
    "alignment_start={} #temporary dictionary\n",
    "alignment_end={} #temporary dictionary\n",
    "\n",
    "#dictionaries collecting output sequences\n",
    "dict_ribozyme={}\n",
    "dict_grey_area={}\n",
    "dict_not_ribozyme={}\n",
    "\n",
    "#collects counts of sequences in each pool\n",
    "alignment_count={}\n",
    "alignment_count['aligned']=0\n",
    "alignment_count['grey_area']=0\n",
    "alignment_count['not_ribozyme']=0\n",
    "alignment_count['plus_aligned']=0\n",
    "alignment_count['minus_aligned']=0\n",
    "alignment_count['plus_and_minus']=0\n",
    "alignment_count['fraction_plus_aligned']=0\n",
    "alignment_count['fraction_minus_aligned'] = 0\n",
    "alignment_count['fraction_plus_and_minus'] =0\n",
    "\n",
    "dict_alignment_noplusandminus = {}\n",
    "\n",
    "list_len_seq =[]\n",
    "list_score=[]\n",
    "list_score_norm=[]\n",
    "classification =[]\n",
    "sequences =[]\n",
    "\n",
    "track =0\n",
    "track2=0\n",
    "    \n",
    "for r in range(len(ribozyme)):\n",
    "    r_name = ribozyme_names[r]\n",
    "    dict_alignment[r_name]={} #add subdictionary\n",
    "    dict_alignment_noplusandminus[r_name]={}    \n",
    "\n",
    "    r_len = len(ribozyme[r]) #length of ribozyme\n",
    "    positions_list = []\n",
    "    for z in range(r_len):\n",
    "        positions_list.append(z) #list of all positions ribozyme\n",
    "            \n",
    "    for p1 in positions_list: #prepare dictionary with all positions --> makes sure only record positions that exist as \n",
    "        #type 1 shorter than t5\n",
    "        dict_alignment[r_name][p1]={}\n",
    "        dict_alignment[r_name][p1]['score']=0\n",
    "        dict_alignment[r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "        dict_alignment_noplusandminus[r_name][p1]={}\n",
    "        dict_alignment_noplusandminus[r_name][p1]['score']=0\n",
    "        dict_alignment_noplusandminus[r_name][p1]['sequence']=ribozyme[r][p1]\n",
    "    \n",
    "    dict_ribozyme[r_name]={}\n",
    "\n",
    "    \n",
    "for seq in dict_random: #go through each file of filtered dictionary\n",
    "    #can insert length limitation here if only want to include length of sequences not filtered out previously\n",
    "    #change cutoff values here:\n",
    "    track2+=1\n",
    "    seq_len = len(seq)\n",
    "    #append seq_len to list_len_seq\n",
    "    list_len_seq.append(seq_len)\n",
    "    sequences.append(seq)\n",
    "\n",
    "    cutoff_ribozyme = 0.66+2/seq_len\n",
    "    cutoff_not_ribozyme = 0.6\n",
    "    \n",
    "    aligned = 0\n",
    "    not_ribozyme = 0\n",
    "    alignment_start[seq]={}\n",
    "    alignment_end[seq]={}\n",
    "    help_dict_align[seq]={}\n",
    "    \n",
    "        \n",
    "    count += dict_random[seq]\n",
    "\n",
    "    for r1 in range(len(ribozyme)):\n",
    "        alignment = pairwise2.align.localms(ribozyme[r1],seq,1,0,-1000,-1000) #1 for match, 0 for mismatch, -1000 for gap\n",
    "        alignment_score = alignment[0][2]\n",
    "        list_score.append(alignment_score)\n",
    "\n",
    "        alignment_score_normalized = alignment_score/seq_len\n",
    "        list_score_norm.append(alignment_score_normalized)\n",
    "\n",
    "        if alignment_score_normalized <cutoff_not_ribozyme:\n",
    "            not_ribozyme +=1\n",
    "\n",
    "        if alignment_score_normalized > cutoff_ribozyme: #cut off adjusted for length --> determined through looking at alignment distributions for different pools\n",
    "            aligned+=1\n",
    "            help_dict_align[seq][ribozyme_names[r1]] = dict_random[seq]\n",
    "            dict_ribozyme[ribozyme_names[r1]][seq]=dict_random[seq]\n",
    "            alignment_start[seq][ribozyme_names[r1]]=alignment[0][3] #retrieves start\n",
    "            alignment_end[seq][ribozyme_names[r1]]=alignment[0][4] #retrieves end\n",
    "\n",
    "    if aligned == 0:\n",
    "        if not_ribozyme == 4:\n",
    "            alignment_count['not_ribozyme']+= dict_random[seq]  \n",
    "            classification.append('not_ribozyme')\n",
    "            dict_not_ribozyme[seq]=dict_random[seq]\n",
    "        else:\n",
    "            alignment_count['grey_area']+=dict_random[seq]\n",
    "            dict_grey_area[seq]=dict_random[seq]\n",
    "            classification.append('grey_area')\n",
    "    else:\n",
    "        alignment_count['aligned']+=dict_random[seq]\n",
    "        classification.append('aligned')\n",
    "\n",
    "    #if track2 divisible by 1000, print track2 and time\n",
    "    if track2%1000 == 0:\n",
    "        print(track2)\n",
    "        print(datetime.datetime.now())\n",
    "        \n",
    "for seq in help_dict_align: #write scores into alignment dictionary \n",
    "    \n",
    "    plus_strand = 0\n",
    "    minus_strand =0\n",
    "    for r2 in range(len(ribozyme)):\n",
    "        if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "            if (ribozyme_names[r2] == 'type5') or (ribozyme_names[r2] == 'type1'):\n",
    "                plus_strand +=1\n",
    "            else:\n",
    "                minus_strand +=1\n",
    "            my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "            for y in range(my_range):\n",
    "                position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "                \n",
    "                s = dict_random[seq]/(count*len(help_dict_align[seq]))\n",
    "                dict_alignment[ribozyme_names[r2]][position]['score']+=s\n",
    "\n",
    "    if (plus_strand >0) and (minus_strand ==0):\n",
    "        alignment_count['plus_aligned'] += dict_random[seq]\n",
    "        #change value in classification list at same position as seq\n",
    "        classification[track]='plus_aligned'\n",
    "\n",
    "    if (plus_strand >0) and (minus_strand >0):\n",
    "        alignment_count['plus_and_minus'] += dict_random[seq]\n",
    "        classification[track]='plus_and_minus'\n",
    "    if (plus_strand ==0) and (minus_strand >0):\n",
    "        alignment_count['minus_aligned'] += dict_random[seq]\n",
    "        classification[track]='minus_aligned'\n",
    "    \n",
    "    if ((plus_strand == 0) and (minus_strand > 0)) or ((plus_strand > 0) and (minus_strand == 0)):\n",
    "        for r2 in range(len(ribozyme)):\n",
    "            if ribozyme_names[r2] in help_dict_align[seq]:\n",
    "                    \n",
    "                my_range = alignment_end[seq][ribozyme_names[r2]] - alignment_start[seq][ribozyme_names[r2]]\n",
    "                for y in range(my_range):\n",
    "                    position = alignment_start[seq][ribozyme_names[r2]]+y #getting position on ribozyme to put into dictionary for each nucleotide in sequence\n",
    "\n",
    "                    s = dict_random[seq]/(count*len(help_dict_align[seq])) #len(help_dict_align[seq] is number of ribozyme sequences aligned to\n",
    "                    dict_alignment_noplusandminus[ribozyme_names[r2]][position]['score']+=s\n",
    "\n",
    "    track +=1\n",
    "    \n",
    "        \n",
    "alignment_count['fraction_plus_aligned']=alignment_count['plus_aligned']/alignment_count['aligned']\n",
    "alignment_count['fraction_minus_aligned'] = alignment_count['minus_aligned']/alignment_count['aligned']\n",
    "alignment_count['fraction_plus_and_minus'] = alignment_count['plus_and_minus']/alignment_count['aligned']\n",
    "\n",
    "\n",
    "\n",
    "#make dataframe with lists sequences, list_len_seq, list_score, list_score_norm, classification as columns\n",
    "list_score_norm_max = []\n",
    "list_score_max=[]\n",
    "for i in range(0, len(list_score_norm), 4):\n",
    "    list_score_norm_max.append(max(list_score_norm[i:i+4]))\n",
    "    list_score_max.append(max(list_score[i:i+4]))\n",
    "\n",
    "df=pd.DataFrame({'sequences':sequences, 'len_seq':list_len_seq, 'score':list_score_max, 'score_norm':list_score_norm_max, 'classification':classification})\n",
    "\n",
    "#make column class where class is classification if \"not_ribozyme\" or \"grey_area\" and is 'aligned' for \"plus_aligned\",  \"minus_aligned\",\"plus_and_minus\"\n",
    "df['class'] = df['classification']\n",
    "df.loc[df['classification']=='plus_aligned', 'class'] = 'aligned'\n",
    "df.loc[df['classification']=='minus_aligned', 'class'] = 'aligned'\n",
    "df.loc[df['classification']=='plus_and_minus', 'class'] = 'aligned'\n",
    "\n",
    "df.to_csv(\"Length_vs_alignment_score_random_HI.csv\",index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2da7c444e7cf019ee465466dbd156869885bf0cbe02a9a56cc171d49857f5cbc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
